File: macos/Sources/BridgeyMac/BridgeyDiagnostics.swift
import Foundation

struct BridgeyDiagnosticEvent: Codable, Equatable {
    let timestamp: String
    let category: String
    let event: String
    let outcome: String
}

final class BridgeyDiagnostics {
    private let limit: Int
    private var events: [BridgeyDiagnosticEvent] = []
    private let formatter = ISO8601DateFormatter()

    init(limit: Int = 200) { self.limit = limit }

    func record(category: String, event: String, outcome: String = "ok") {
        events.append(BridgeyDiagnosticEvent(
            timestamp: formatter.string(from: Date()),
            category: token(category),
            event: token(event),
            outcome: token(outcome)
        ))
        if events.count > limit { events.removeFirst(events.count - limit) }
    }

    func snapshot() -> [BridgeyDiagnosticEvent] { events }

    func report(
        connectionState: String,
        transfers: [FileTransferRow],
        localFeatures: [BridgeyFeature: Bool],
        remoteFeatures: [BridgeyFeature: Bool]
    ) throws -> Data {
        let featureObject: ([BridgeyFeature: Bool]) -> [String: Bool] = { features in
            Dictionary(uniqueKeysWithValues: BridgeyFeature.allCases.map { ($0.rawValue, features[$0] != false) })
        }
        let payload: [String: Any] = [
            "schemaVersion": 1,
            "generatedAt": formatter.string(from: Date()),
            "appVersion": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "development",
            "platform": "macos",
            "osVersion": ProcessInfo.processInfo.operatingSystemVersionString,
            "connectionState": token(connectionState),
            "activeTransferCount": transfers.filter { $0.active }.count,
            "historyCount": transfers.filter { !$0.active }.count,
            "localFeatures": featureObject(localFeatures),
            "remoteFeatures": featureObject(remoteFeatures),
            "events": try events.map { event in
                let data = try JSONEncoder().encode(event)
                return try JSONSerialization.jsonObject(with: data)
            },
        ]
        return try JSONSerialization.data(withJSONObject: payload, options: [.prettyPrinted, .sortedKeys])
    }

    private func token(_ value: String) -> String {
        String(value.lowercased().map { character in
            character.isASCII && (character.isLetter || character.isNumber || "_.-".contains(character)) ? character : "_"
        }.prefix(64))
    }
}

File: android/app/src/main/java/dev/bridgey/android/BridgeyDiagnostics.kt
package dev.bridgey.android

import android.content.Context
import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant

internal data class DiagnosticEvent(
    val timestamp: String,
    val category: String,
    val event: String,
    val outcome: String,
)

internal class BridgeyDiagnostics(private val limit: Int = 200) {
    private val events = ArrayDeque<DiagnosticEvent>()

    @Synchronized fun record(category: String, event: String, outcome: String = "ok") {
        events.addLast(DiagnosticEvent(Instant.now().toString(), token(category), token(event), token(outcome)))
        while (events.size > limit) events.removeFirst()
    }

    @Synchronized fun snapshot(): List<DiagnosticEvent> = events.toList()

    fun report(
        context: Context,
        connectionState: String,
        transfers: Collection<FileTransferState>,
        localFeatures: Map<BridgeyFeature, Boolean>,
        remoteFeatures: Map<BridgeyFeature, Boolean>,
    ): String {
        val version = runCatching {
            context.packageManager.getPackageInfo(context.packageName, 0).versionName
        }.getOrNull() ?: "unknown"
        return JSONObject().apply {
            put("schemaVersion", 1)
            put("generatedAt", Instant.now().toString())
            put("appVersion", version)
            put("platform", "android")
            put("osApi", Build.VERSION.SDK_INT)
            put("connectionState", token(connectionState))
            put("activeTransferCount", transfers.count(FileTransferState::active))
            put("historyCount", transfers.count { !it.active })
            put("localFeatures", featureObject(localFeatures))
            put("remoteFeatures", featureObject(remoteFeatures))
            put("events", JSONArray().apply {
                snapshot().forEach { event ->
                    put(JSONObject().apply {
                        put("timestamp", event.timestamp)
                        put("category", event.category)
                        put("event", event.event)
                        put("outcome", event.outcome)
                    })
                }
            })
        }.toString(2)
    }

    private fun featureObject(features: Map<BridgeyFeature, Boolean>) = JSONObject().apply {
        BridgeyFeature.entries.forEach { put(it.key, features[it] != false) }
    }

    private fun token(value: String): String = value.lowercase()
        .replace(Regex("[^a-z0-9_.-]"), "_")
        .take(64)
}
